### PR TITLE
Add in-place conversion to basic data format

### PR DESCRIPTION
**Pull request by mtanneau**
Motivated by #913 

This PR introduces an in-place `make_basic_data!`, which has the same functionality but does not create a `deepcopy`.
The goal is to avoid the (time and memory) cost of deepcopy-ing the initial data dictionary when the user does not want to keep it.

A typical use-case would be workflows that look like `data = make_basic_network(parse_file(...))`.
This call with load the raw data dictionary with `parse_file`, then deep-copy it, convert the copy to basic format, and discard the original raw data.

Timings of `make_basic_network` after the changes: times are improved, and memory consumption is reduced.
Note: those are just the ouptut of `@time`; I did run `GC.gc()` before to reduce garbage collection, but it's not a full-fledge `@benchmark`

| Case | #buses | old time (s) | in-place time (s) | old memory (MiB) | in-place memory (MiB) |
|:--|--:|--:|--:|--:|--:|
| `1354_pegase` | 1354 | 0.21 | 0.13 | 29.4 | 20.4 |
| `2869_pegase` | 2869 | 0.60 | 0.40 | 66.7 | 48.2 |
| `9241_pegase` | 9241 | 2.50 | 1.87 | 208.3 | 145.9 |
| `13659_pegase` | 13659 | 5.44 | 4.67 | 313.0 | 227.4 |

and corresponding profile of `make_basic_network!(data)` where `data = pglib("9241_pegase")`:
prof_inplace

Note that the above profile does not include the improvements in #914. The main (and expected) outcome is that the time spent deep-copying at the beginning is gone.

### CODE REVIEW

**Comment by ccoffrin** [comment]
Shall we add a test for `make_basic_network!` and `make_basic_network` that confirms that one is destructive and the other is not?

**Comment by mtanneau** [comment]
I can do that. I'm currently in a plane, will push in a couple days

**Comment by ccoffrin** [comment]
After that I'll be good to merge.  Feel free to add your name in the list of the readme if you like.

File: test/data-basic.jl [hunk 1]
```diff
@@ -35,6 +35,35 @@
...
     end
 
+    @testset "In-place make_basic_network!" begin
+        data = PowerModels.parse_file("../test/data/matpower/case7_tplgy.m")
+        data_ = deepcopy(data)
+        
+        # Test #1: `make_basic_network` is non-destructive
+        data_basic = PowerModels.make_basic_network(data)
+        @test data == data_        # input data not modified
+        @test data_basic !== data  # we created a new dict
```
Comment (mtanneau) [inline test/data-basic.jl:45]: These ones check that we did not modify the input data, and that we returned a new `Dict` object

File: test/data-basic.jl [hunk 2]
```diff
@@ -35,6 +35,35 @@
...
+
+        # Test #2: `make_basic_network!` is destructive and works in-place
+        data_basic = PowerModels.make_basic_network!(data)
+        @test data_copy === data  # in-place modification
```
Comment (mtanneau) [inline test/data-basic.jl:49]: Here I only check that the returned dictionary is exactly the one that was passed as input.
Note the use of triple equal `===`

**Pull request merged**